Add lock to kernel startup

ajbozarth opened the pull request:
Currently if you try to start a new run while the kernel is starting
it wint stop the previous run and the two runs output will be spliced

This adds a "lock" around starting the kernel, the lock prevents
users from starting a new run by both disabling both the run button
and the `runPython()` function if statement.

Fixes #394
Fixes #398


 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

kevin-bates commented on packages/python-runner/src/PythonRunner.ts:
```diff
@@ -18,35 +18,49 @@ import { Kernel } from '@jupyterlab/services';
       try {
+        this.startKernelLock(true);
         this.kernel = await this.startKernel(kernelSettings);
+        this.startKernelLock(false);
```
I think it might be better to keep the lock enabled for the duration of the run or until the 'stop' button is selected.  I don't really see any reason to enable the run button when the kernel is already running and its enablement just makes it easier to disrupt things.

Unrelated but thought I'd raise here...
I would recommend using unique text (however slight it is) in error messages so that the location of where the error occurred can be identified since what we're usually provided is a description or screenshot at best.  A couple lines below we have duplicate messages:  'Could not start kernel environment to execute script.'  I suggest changing one to differentiate it from the other.

ajbozarth replied:
the reason the lock is only implemented here is because my previous PR added support for interrupting runs by starting a new run. The edge case this is addressing is just more common with that new feature.

kevin-bates commented:
I had this occur today when selecting the run button during a run.  This happened multiple times, but not always.  I had two runners going and it seemed to only happen on the second ('b') of the two. Will try with a single python runner.
![image](https://user-images.githubusercontent.com/22599560/78381261-f7685280-7589-11ea-9101-a4d9e1abe154.png)

Here's the code in each runner - with the character changed to 'a' in the other to differentiate:
```python
import time

i = 0
while i < 60:
    print('b: ' + str(i))
    time.sleep(1)
    i = i+1
print('b: ' + str(i))
```

ajbozarth commented:
Based on discussion at today's scrum I will be doing the following on this PR:
- Lock on the whole run, this mean a user can only run again after clicking stop or letting it finish
- Move handling of stopKernel introduced in #392 to the close tab callback only
- if appropriate, address #398 while updating how stops are handled.
- if appropriate address the separate error text issue raise in @kevin-bates comment

karlaspuldaro commented on packages/python-runner/src/PythonRunner.ts:
```diff
@@ -18,35 +18,49 @@ import { Kernel } from '@jupyterlab/services';
+    this.startingKernel = false;
   }
 
+  startKernelLock(lock: boolean): void {
```
I don't think DOM manipulation on an element defined in the widget should be done in PythonRunner, since it's supposed to only deal with the backend. I was thinking some sort of callback like we did with `handleKernelMsg()` that notifies the UI, so that button lock is done from the widget.

ajbozarth replied:
That's fair, I moved it in here because I needed to access it from multiple functions, but I realized I would hand it to the constructor as a callback function instead of passing in the id and it would work the same.

ajbozarth commented:
I've pushed an update that addresses the following:
- Lock on the whole run
- Move function that handles locking to widget and passes the function to the PythonRunner
- Created a helper function for handling error dialogs that makes sure to unlock on error
- Added a call to stopKernel for error case where the kernel exists but is in an unusable state
- Make a stopRun handler that removes the `[*]` on stop (conditional on the output being visible to prevent undefined errors in the log)
- Added a lock flag in addition to disabling the button since a disabled button does not prevent programmatic calls to handlers, meaning without the flag and a check for it in runPython the button would still work even though it's "disabled"

Note there is one odd edge case remaining with this fix, but it has a simple and obvious workaround and thus isn't worth addressing. If a user closes the output tab while the kernel is starting it will sometime not stop the kernel and the run will continue "in the background" leaving the run button disabled until it finishes. The obvious workaround for this is to just click the stop button, which then stops the "background" run and reenables the run button.
